### DISCUSSION TITLE
Player-facing editors

### DISCUSSION THREAD

**Discussion by B-Reif** [bevyengine/bevy#2141] [Ideas]
Many games benefit from exposing a GUI editor to their players, allowing their communities to create custom content. Typically, these editors derive from the authoring tools used by the game developers themselves. By designing Bevy's editor with support for a "player-facing" side, we could enable more games to release player-facing tools.

Considerations:

- If player-facing tools can inject compiled code, that coding environment must be "sandboxed" to prevent arbitrary interaction with the engine and game code. On Bevy's side, this means giving game devs tools to design and expose their own editor-friendly coding APIs.
- Player-facing tools should create serializable assets in a way that's purely additive. In other words, players should not be able to destructively change the content of their game client. (Player-made assets could still override existing ones during runtime.)
- Games which support a player-facing editor would ideally be able to "dogfood" the editor themselves to make their game content, with some of their features protected behind a developer flag. For example, Blizzard's RTS editors have "Blizzard-signed" functions reserved for official content (such as a function for unlocking an achievement when something happens).

**Comment by alice-i-cecile** [comment]
I like this idea a lot. This feature was *fantastic* in WC3 for me when I was young.

I don't know that there's a lot actionable for it *right* now, but this should be borne in mind as we build out the editor.
